src/Stratis.FederatedPeg.Features.FederationGateway/SourceChain/MaturedBlocksProvider.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Features.BlockStore;
+using Stratis.FederatedPeg.Features.FederationGateway.Interfaces;
+
+namespace Stratis.FederatedPeg.Features.FederationGateway.SourceChain
+{
+    public class MaturedBlocksProvider : IMaturedBlocksProvider
+    {
+        private ConcurrentChain chain;
+        private IDepositExtractor depositExtractor;
+        private IBlockRepository blockRepository;
+
+        public MaturedBlocksProvider(ILoggerFactory loggerFactory,
+            ConcurrentChain chain, IDepositExtractor depositExtractor, IBlockRepository blockRepository)
+        {
+            this.chain = chain;
+            this.depositExtractor = depositExtractor;
+            this.blockRepository = blockRepository;
+        }
+
+        /// <inheritdoc />
+        public async Task<List<IMaturedBlockDeposits>> GetMaturedDepositsAsync(int blockHeight, int maxBlocks)
+        {
+            int currentHeight = blockHeight;
+            int matureHeight = (this.chain.Tip.Height - (int)this.depositExtractor.MinimumDepositConfirmations);
+
+            if (currentHeight > matureHeight)
+            {
+                throw new Exception($"Block height {currentHeight} submitted is not mature enough. Blocks less than a height of {matureHeight} can be processed.");
+            }
+
+            // Pre-load the block data
+            var blockHashes = new List<uint256>();
+            var chainedHeaders = new ChainedHeader[Math.Min(maxBlocks, matureHeight - currentHeight)];
```

Comment:
Could there be a potential race condition if this method was invoked by separate threads ?

Reply:
@ferdeen, I'm not seeing it. Can you provide more information?